### DISCUSSION TITLE
Apollo 3 caching support

### DISCUSSION THREAD

**Discussion by art049** [MichalLytek/type-graphql#1371] [Q&A]
Hello, thanks a lot for this amazing package!

I'm currently looking to implement server side caching with Apollo 3 and I didn't find a lot of references.
Any ideas or sample implementations I could use to do it?

**Few references of related things I found out here:**

- [Can i get some guide to use cacheControl with ApolloServer ?](https://github.com/MichalLytek/type-graphql/discussions/919#discussioncomment-1141918)
- [apollo-cache example](https://github.com/MichalLytek/type-graphql/tree/master/examples/apollo-cache) doesn't work anymore since `apollo-cache-control` has been deprecated.
- https://github.com/MichalLytek/type-graphql/pull/369 : Hence it's possible to use `@Directive("@cacheControl(maxAge: 60, scope: PUBLIC)")` for example. However, Apollo doesn't catch the directives.

**Are you able to make a PR that fix this?**
I'd love to do that but I didn't find any solution to make it work yet.

**Comment by okaforcj** [comment]
@MichalLytek please have you seen this. it wowuld be much appreciated to give any advice on this

**Reply by MichalLytek** [reply]
What have you tried and what was not working for you?

**Reply by okaforcj** [reply]
thanks for the quick reply @MichalLytek  the only documentation on something similar is here : https://github.com/MichalLytek/type-graphql/pull/369 and it seems to be outdated, I'm using apollo 4 but similar to apollo 3 in that all caching is represented via apollo-server-core. as per apollo's docs:
"To use the @ cacheControl directive, you must add the following definitions to your server's schema:

```graphql
enum CacheControlScope {
  PUBLIC
  PRIVATE
}

directive @cacheControl(
  maxAge: Int
  scope: CacheControlScope
  inheritMaxAge: Boolean
) on FIELD_DEFINITION | OBJECT | INTERFACE | UNION
```

it keeps error out with unknown directive "@ cacheControl"

after i try to use

```js
@Directive("@cacheControl(maxAge: 30)")
```

**Reply by MichalLytek** [reply]
You need to declare the directive too by `new GraphQLDirective` and provide it in build options

**Reply by okaforcj** [reply]
thanks for thiis, please can you give me an example of this or point to where the documentation of the graphqldirective is used because i dont see it for type-graphql. i started like this but im not sure its correct or how to implement the apollo directive within it 
``` js
    const schema = await buildSchema({
        resolvers: [__dirname + "/resolvers/**/*.{ts,js}"],
        emitSchemaFile: path.resolve(__dirname, "../app/lib/graphql/schema.graphql"),
        validate: true,
        directives:[ new GraphQLDirective(...)]
    })
```

**Reply by art049** [reply]
> You need to declare the directive too by new GraphQLDirective and provide it in build options

I tried to do this @MichalLytek but unfortunately couldn't get this to work neither.

**Reply by MichalLytek** [reply]
What do you have exactly inside the `new GraphQLDirective(...)`?

**Reply by art049** [reply]
I tried with this directive added to the schema
```js
 new GraphQLDirective({
  name: "cacheControl",
  locations: ["FIELD_DEFINITION", "OBJECT", "INTERFACE", "UNION"],
  args: {
    maxAge: {
      type: GraphQLInt,
    },
    scope: {
      type: new GraphQLEnumType({
        name: "CacheControlScope",
        values: {
          PUBLIC: { value: "PUBLIC" },
          PRIVATE: { value: "PRIVATE" },
        },
      }),
    },
    inheritMaxAge: {
      type: GraphQLBoolean,
    },
  },
});
```

The cache ends up working on root queries. But doesn't on field resolvers which is a big issue on my side. 
Do you have any idea @MichalLytek ?
